R/add_cols.R

```diff
@@ -0,0 +1,293 @@
+#' Add event date as column to infectious history `<data.frame>`
+#'
+#' @description The event date could be first contact, last contact or other.
+#'
+#' @param .data A `<data.frame>` containing the infectious history from a
+#' branching process simulation.
+#' @param contact_type A `character` with the type of contact, either first
+#' contact (`"first"`), or last contact (`"last"`).
+#' @param distribution A `character` with the name of the distribution,
+#' following the base R convention for distribution naming (e.g. Poisson
+#' is `pois`).
+#' @param ... [dots] Extra arguments to be passed to the distribution function
+#' given in the `distribution` argument.
+#' @inheritParams sim_linelist
+#'
+#' @name .add_date
+#'
+#' @return A `<data.frame>` with one more column than input into `.data`.
+#' Unless the column heading is already present in which the data is
+#' overwritten.
+#' @keywords internal
+#'
+NULL
+
+#' @name .add_date
+.add_date_contact <- function(.data,
+                              contact_type = c("first", "last"),
+                              distribution = c("pois", "geom"),
+                              ...,
+                              outbreak_start_date = NULL) {
+  contact_type <- match.arg(contact_type)
+  distribution <- match.arg(distribution)
+
+  stopifnot(
+    "outbreak_start_date is only required for adding date of last contact" =
+      contact_type == "last" && !is.null(outbreak_start_date) ||
+      contact_type == "first" && is.null(outbreak_start_date)
+  )
+
+  rdist <- switch(distribution,
+    pois = stats::rpois,
+    geom = stats::rgeom
+  )
+
+  # c() over ...length() to ensure NULL is not counted by length
+  if (length(c(...)) == 0) {
+    stop("Distribution parameters are missing, check config", call. = FALSE)
+  }
```

Comment:
There is a mixture of `stopifnot()` and `stop()` used throughout the package. It might be worth at some point looking through and seeing if `stopifnot()` can be used in all cases. I'm not sure that the behaviour of `stop(..., call. = FALSE)` can be replicated in `stopifnot()`.